Define IEquatable when already implemented

ie-zero opened the pull request:
All the classes amended have had implemented the IEquatable<T> interface already. This change simply add the interface at the class definition.

ie-zero committed:
Define IEquatable when implemented [v2]

All the classes amended have had implemented the
IEquatable<T> interface already. This change simply add the interface at
the class definition.

The changes are affecting only the v2 project.

ie-zero committed:
Define IEquatable when implemented [.NET Core]

All the classes amended have had implemented the
IEquatable<T> interface already. This change simply add the interface at
the class definition.

The changes are affecting only the .NET Core project.

ie-zero committed:
Define IComparable when already implemented

Add IComparable and IComparable<T> interface definitions to classes
which already implement the funcitonality.

The changes are affecting both v2 and .NET Core projects.

rianjs commented on net-core/Ical.Net/Ical.Net/DataTypes/PeriodList.cs:
```diff
@@ -30,7 +31,7 @@ public PeriodList(string value) : this()
-        protected bool Equals(PeriodList other)
+        public bool Equals(PeriodList other)
         {
             return string.Equals(TzId, other.TzId) && CollectionHelpers.Equals(Periods, other.Periods);
```
While you're here, let's make all `string.Equals` use `StringComparer.OrdinalIgnoreCase`.

rianjs commented on net-core/Ical.Net/Ical.Net/DataTypes/RequestStatus.cs:
```diff
@@ -62,7 +63,7 @@ public override string ToString()
+        public bool Equals(RequestStatus other)
         {
             return string.Equals(_mDescription, other._mDescription) && string.Equals(_mExtraData, other._mExtraData) &&
                    Equals(_mStatusCode, other._mStatusCode);
```
Let's use `OrdinalIgnoreCase`, here and elsewhere.

rianjs commented on net-core/Ical.Net/Ical.Net/DataTypes/WeekDay.cs:
```diff
@@ -64,6 +66,20 @@ public override void CopyFrom(ICopyable obj)
 
+        public int CompareTo(WeekDay other)
+        {
+            if (other == null)
```
Throwing an exception in a `CompareTo` method probably isn't a good idea. Consider the usage in a LINQ query:

```csharp
var orderedDays = weekdayCollection
    .OrderBy(w => w)
    .ToList();
```

That could `throw` in the event that WeekDay is null. I think being able to sort `null` is reasonable behavior: any non-null WeekDay has greater weight. Consider instead:

```csharp
public int CompareTo(WeekDay other)
{
    if (ReferenceEquals(this, other)) return 0;
    return ReferenceEquals(null, other)
        ? 1
        : DayOfWeek.CompareTo(other.DayOfWeek);
}
```

I'm not sure using `Offset` as a tie-breaker is necessary or desirable.

rianjs commented on net-core/Ical.Net/Ical.Net/DataTypes/WeekDay.cs:
```diff
@@ -64,6 +66,20 @@ public override void CopyFrom(ICopyable obj)
+            return compare;
+        }
+
         public int CompareTo(object obj)
```
Let's get rid of this `IComparable` implementation in favor of just `IComparable<T>`, so:
* Delete this method
* Delete `IComparable` from `IWeekDay` in v2, and delete `IComparable` from the top here in net-core.

ie-zero committed:
Enforce case-insensitive string comparison on Equal()

The case-insensitive string comparison is enformced on implementations
of Equal() method as the RFC 5545 does not specify, for properties and
most other component values, case sensitivity in most cases.

ie-zero committed:
Reimplement IComparable and IComparable<T>

* Remove IComparable (non-generic) implementation of the interface
whenever possible.
* Re-implement all CompareTo(T) methods so they do not throw exceptions
and unify the implementation across all methods.

rianjs commented on net-core/Ical.Net/Ical.Net/Components/UniqueComponent.cs:
```diff
@@ -96,8 +96,28 @@ protected override void OnDeserialized(StreamingContext context)
+                    return Uid.Equals(r.Uid);
+                }
+                return Uid == r.Uid;
+            }
```
* Using the `Uid` as the basis for equality isn't consistent with the rest of the library. (I know that's the current implementation -- which is wrong, so not your fault.)
* This method should delegate to `RecurringComponent.Equals()` in the event that the element is recurring, and call its base class in the event that the element is NOT recurring. I don't think there's anything inside `UniqueComponent` that's useful for determining equality or hashing. 

`RecurringComponent` should know how to do a reasonable comparison and hash, but I think it's missing the `IEquatable<T>`, and isn't defining an `Equals` method, so we have nothing to delegate to. That will need to be implemented.

If you want, you can add the `IEquatable<T>` to the `RecurringComponent`, and define an `Equals` function. I think it will require some thought for both `GetHashCode` and `Equals`. If you don't want to do that, or would prefer to do it as a separate pull request, you can remove the `IEquatable` from this class, and add it back in with the changes to RecurringComponent. (Or I'll do it myself.)

Actually, I think that work should be done as a separate ticket, because it's a little more significant, and needs to be done carefully.

rianjs commented on v2/ical.NET/Components/UniqueComponent.cs:
```diff
@@ -95,8 +95,26 @@ protected override void OnDeserialized(StreamingContext context)
+            }
+            return base.Equals(other);
+        }
+
```
(Same comment here,)

rianjs reviewed (approved):
Can you merge this after I submit my approval, or do I need to tweak the repo settings?

ie-zero committed:
Increase minor version and update release notes
